File: hmode/shared/libs/vectorstore-unified/typescript/src/embeddings/cohere.ts
/**
 * Cohere embeddings provider (direct API).
 */
import { BaseEmbeddings } from './base.js';

interface CohereEmbeddingsOptions {
  model?: string;
  apiKey?: string;
}

const MODEL_DIMENSIONS: Record<string, number> = {
  'embed-english-v3.0': 1024,
  'embed-multilingual-v3.0': 1024,
  'embed-english-light-v3.0': 384,
};

export class CohereEmbeddings extends BaseEmbeddings {
  readonly dimension: number;
  private readonly model: string;
  private readonly apiKey: string;

  constructor(options: CohereEmbeddingsOptions = {}) {
    super();
    this.model = options.model ?? 'embed-english-v3.0';
    this.dimension = MODEL_DIMENSIONS[this.model] ?? 1024;
    this.apiKey = options.apiKey ?? process.env.COHERE_API_KEY ?? '';

    if (!this.apiKey) {
      throw new Error('Cohere API key required');
    }
  }

  async embedText(text: string): Promise<number[]> {
    const embeddings = await this.callCohere([text], 'search_document');
    return embeddings[0];
  }

  async embedTexts(texts: string[]): Promise<number[][]> {
    return this.callCohere(texts, 'search_document');
  }

  async embedQuery(query: string): Promise<number[]> {
    const embeddings = await this.callCohere([query], 'search_query');
    return embeddings[0];
  }

  private async callCohere(
    texts: string[],
    inputType: string
  ): Promise<number[][]> {
    const response = await fetch('https://api.cohere.ai/v1/embed', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        Authorization: `Bearer ${this.apiKey}`,
      },
      body: JSON.stringify({
        texts,
        model: this.model,
        input_type: inputType,
      }),
    });

    if (!response.ok) {
      throw new Error(`Cohere API error: ${response.statusText}`);
    }

    const result = await response.json();
    return result.embeddings;
  }
}

File: hmode/shared/libs/vectorstore-unified/python/vectorstore_unified/embeddings/cohere.py
"""Cohere embeddings provider (direct API)."""

import os
from typing import List, Optional

from .base import BaseEmbeddings


class CohereEmbeddings(BaseEmbeddings):
    """
    Cohere embeddings via direct API.

    Models:
        - embed-english-v3.0 (1024 dims)
        - embed-multilingual-v3.0 (1024 dims)
        - embed-english-light-v3.0 (384 dims, faster)
    """

    MODEL_DIMENSIONS = {
        "embed-english-v3.0": 1024,
        "embed-multilingual-v3.0": 1024,
        "embed-english-light-v3.0": 384,
    }

    def __init__(
        self,
        model: str = "embed-english-v3.0",
        api_key: Optional[str] = None,
    ):
        """
        Initialize Cohere embeddings.

        Args:
            model: Cohere model name
            api_key: Cohere API key (defaults to COHERE_API_KEY env var)
        """
        try:
            import cohere
        except ImportError:
            raise ImportError("cohere package required: pip install cohere")

        self.model = model
        self._dimension = self.MODEL_DIMENSIONS.get(model, 1024)

        api_key = api_key or os.environ.get("COHERE_API_KEY")
        if not api_key:
            raise ValueError("Cohere API key required")

        self.client = cohere.Client(api_key)

    @property
    def dimension(self) -> int:
        return self._dimension

    def embed_text(self, text: str) -> List[float]:
        """Embed a single text string."""
        response = self.client.embed(
            texts=[text],
            model=self.model,
            input_type="search_document",
        )
        return response.embeddings[0]

    def embed_texts(self, texts: List[str]) -> List[List[float]]:
        """Embed multiple texts."""
        response = self.client.embed(
            texts=texts,
            model=self.model,
            input_type="search_document",
        )
        return response.embeddings

    def embed_query(self, query: str) -> List[float]:
        """Embed a search query."""
        response = self.client.embed(
            texts=[query],
            model=self.model,
            input_type="search_query",
        )
        return response.embeddings[0]
